File: problems/g/guessthedatastructure/guessthedatastructure.py
#! /usr/bin/python3
"""
Author: Rudi César Comiotto Modena
Email: rudi.modena@gmail.com
"""

class Stack:
    def __init__(self):
        self.stack = list()
        
    def push(self, value):
        self.stack.append(value)
        
    def pop(self):
        if self.stack:
            return self.stack.pop()
        else:
            return -1


class Queue:
    def __init__(self):
        self.queue = list()
    
    def push(self, value):
        self.queue.append(value)
    
    def pop(self):
        if self.queue:
            return self.queue.pop(0)
        else:
            return -1


class Priority_Queue:
    def __init__(self):
        self.priority_queue = list()

    def push(self, value):
        self.priority_queue.append(value)
    
    def pop(self):
        if self.priority_queue:
            max_value = max(self.priority_queue)
            self.priority_queue.remove(max_value)
            return max_value
        else:
            return -1


def main():
    while True:
        try:
            n = int(input())
        except EOFError:
            break

        stack = Stack()
        list_stack = list()

        queue = Queue()
        list_queue = list()

        priority_queue = Priority_Queue()
        list_priority_queue = list()

        for _ in range(n):
            command, x = [int(val) for val in input().split()]

            if command == 1:
                stack.push(x)
                queue.push(x)
                priority_queue.push(x)
            
            else:
                if stack.pop() == x:
                    list_stack.append(True)
                else:
                    list_stack.append(False)

                if queue.pop() == x:
                    list_queue.append(True)
                else:
                    list_queue.append(False)

                if priority_queue.pop() == x:
                    list_priority_queue.append(True)
                else:
                    list_priority_queue.append(False)
        
        data_structures = list()
        if all(list_stack):
            data_structures.append("stack")

        if all(list_queue):
            data_structures.append("queue")

        if all(list_priority_queue):
            data_structures.append("priority queue")

        if len(data_structures) == 0:
            print("impossible")
        elif len(data_structures) == 1:
            print(data_structures[0])
        else:
            print("not sure")


if __name__ == '__main__':
    main()

File: problems/g/guessthedatastructure/guessthedatastructure.cpp
/*
Author: Rudi César Comiotto Modena
Email: rudi.modena@gmail.com
*/

#include <bits/stdc++.h>

using namespace std;

int main() {
    int n;
    while (cin >> n) {
        cin.ignore();

        stack<int> mock_stack;
        queue<int> mock_queue;
        priority_queue<int> mock_priority_queue;

        vector<int> valid_stack;
        vector<int> valid_queue;
        vector<int> valid_priority_queue;

        int command, x;
        for (int i=0; i<n; i++) {
            cin >> command >> x;
            cin.ignore();

            if (command == 1) {
                mock_stack.push(x);
                mock_queue.push(x);
                mock_priority_queue.push(x);
            }
            else {
                if (!mock_stack.empty()) {
                    if (mock_stack.top() == x) {
                        valid_stack.push_back(1);
                    }
                    else {
                        valid_stack.push_back(0);
                    }
                    mock_stack.pop();
                }
                else {
                    valid_stack.push_back(0);
                }

                if (!mock_queue.empty()) {
                    if (mock_queue.front() == x) {
                        valid_queue.push_back(1);
                    }
                    else {
                        valid_queue.push_back(0);
                    }
                    mock_queue.pop();
                }
                else {
                    valid_queue.push_back(0);
                }

                if (!mock_priority_queue.empty()) {                    
                    if (mock_priority_queue.top() == x) {
                        valid_priority_queue.push_back(1);
                    }
                    else {
                        valid_priority_queue.push_back(0);
                    }
                    mock_priority_queue.pop();
                }
                else {
                    valid_priority_queue.push_back(0);
                }
            }
        }

        vector<string> results;

        size_t sum_stack = 0;
        for (auto& value : valid_stack) {
            sum_stack += value;
        }
        if (valid_stack.empty() || (sum_stack>0 && sum_stack == valid_stack.size())) {
            results.push_back("stack");
        }

        size_t sum_queue = 0;
        for (auto& value : valid_queue) {
            sum_queue += value;
        }
        if (valid_queue.empty() || (sum_queue>0 && sum_queue == valid_queue.size())) {
            results.push_back("queue");
        }

        size_t sum_priority_queue = 0;
        for (auto& value : valid_priority_queue) {
            sum_priority_queue += value;
        }
        if (valid_priority_queue.empty() || (sum_priority_queue>0 && sum_priority_queue == valid_priority_queue.size())) {
            results.push_back("priority queue");
        }
        
        if (results.size() == 0) {
            cout << "impossible" << endl;
        }
        else if (results.size() == 1) {
            cout << results[0] << endl;
        }
        else {
            cout << "not sure" << endl;
        }
    }
}